templates/tapestry.js

```diff
@@ -1071,11 +1026,104 @@ function tapestryTool(config){
             });
     }
 
+    function renderTooltips() {
+        nodes
+            .filter(d => !d.accessible)
+            .append("foreignObject")
+            .attr("class", "tooltip-wrapper")
+            .style("position", "relative")
+            .style("pointer-events", "none")
+            .style("opacity", 0)
+            .attr("width", d => Math.min(getRadius(d) * 2 + 48, 300))
+            .attr("height", d => getRadius(d) * 2)
+            .attr("x", d => -(Math.min(getRadius(d) * 2 + 48, 300) / 2))
+            .attr("y", d => -(getRadius(d) * 3 + 27.5 + 8))
+            .append("xhtml:div")
+            .attr("class", "tapestry-tooltip")
+            .html(getTooltipHtml)
+    }
+
+    function setNodeListeners(nodes) {
+
+        // Create tooltip for all locked nodes
+        renderTooltips();
+
+        // Remove (potentially) old listeners
+        nodes.on("mouseover", null).on("mouseout", null).on("mouseleave", null);
+
+        nodes.on("mouseover", function (thisNode) {
+
+            // Place this node at the end of the svg so that it appears on top
+            $(this).insertAfter($(this).parent().children().last())
+
+            // Mark this node as the node to link to (potentially)
+            if (linkToDragStarted) {
+                linkToNode = thisNode;
+            }
+        }).on("mouseout", function () {
+
+            // Unmark this node as the node to link to
+            if (linkToDragStarted) {
+                linkToNode = undefined;
+            }
+        });
+
+        nodes
+            .filter(d => !d.accessible)
+            .on("mouseover", function (d) {
+
+                // Place this node at the end of the svg so that it appears on top
+                $(this).insertAfter($(this).parent().children().last())
+
+                // Show unlock conditions tooltip
+                if (d.nodeType !== "grandchild") {
+                    const wrapper = this.querySelector(".tooltip-wrapper");
+                    wrapper.style.opacity = 1;
+                }
+            })
+            .on("mouseleave", function () {
+
+                // Hide unlock conditions tooltip
+                const wrapper = this.querySelector(".tooltip-wrapper");
+                wrapper.style.opacity = 0;
+            })
+    }
+
+    function getTooltipHtml(node) {
+        const str = "To unlock this node: <br />";
+        const wrapper = document.createElement("ul");
+
+        if (node.conditions.length === 0) {
+            const listItem = document.createElement("li");
+            listItem.innerText = "Complete this parent.";
+            wrapper.appendChild(listItem);
+        } else {
+            node.conditions.forEach(cond => {
+                if (!cond.fulfilled) {
+                    const listItem = document.createElement("li");
+                    switch (cond.type) {
+                        case conditionTypes.NODE_COMPLETED: {
+                            const node = getNodeById(cond.value);
```

Comment:
Why is the node ID called "Value"? Should it not be called "Node ID" so it's more clear what it is?

Reply:
@ItsAdel can you look at the above bug? I think this was the thing that we discovered a while back. I thought it was fixed.